### PR TITLE
Documentation on setting up a development environment

### LINKED ISSUE #1543: Docs page "Setting up your development environment"
See #1426

### PR DESCRIPTION
**Pull request by haydenshively**
This should close #1543

### CODE REVIEW

File: docs/user/setting_up_your_development_environment.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#Setting up your development environment
```
Comment (irisliucy) [inline docs/user/setting_up_your_development_environment.md:1]: `# Setting up your development environment. `

There should be a space after every header sign for markdown. Same for the others.

**Comment by irisliucy** [comment]
Overall, the doc looks good!

File: docs/user/setting_up_your_development_environment.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# Setting up your development environment
+## Overview
+Garage is a Python package which can be installed in most Python 3.6+
+environments using standard commands, i.e.
+
+``pip install --user garage``
+
+However, to ensure that your environment is consistent with other
```
Comment (avnishn) [inline docs/user/setting_up_your_development_environment.md:8]: We use these tools because they guarantee that there are no packages interfering with the required packages to run garage. This also ensures that a garage installation does not interfere with other packages installed on a user's machine.

File: docs/user/setting_up_your_development_environment.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+- Linux:
+```
+./scripts/setup_linux.sh --mjkey path-to-your-mjkey.txt
+```
+- macOS 10.x:
+
+A list of verified macOS versions can be found at the top of the script. Most
+versions of macOS 10 should work, but we can't speak for versions not in the
+list. It is unlikely that Garage and its dependencies will work out-of-the-box
+with newer Mac's that use Apple's ARM processors.
```
[the remark this answers is not in the document]
Comment (zequnyu) [reply docs/user/setting_up_your_development_environment.md:44]: lol that's quick though.

**Comment by ryanjulian** [comment]
Don't forget to add an acknowledgement for yourself at the bottom of the page! See https://github.com/rlworkgroup/garage/pull/1638 for an example

File: docs/user/setting_up_your_development_environment.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
+# Setting up your development environment
+
+## Overview
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:3]: I think the following 3 sections can be covered by just linking to the installation tutorial which already exists.

File: docs/user/setting_up_your_development_environment.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+Also note that this setup script relies on homebrew to install dependencies.
+If you don't already have homebrew, the script will attempt to install it. If
+you do have homebrew, make sure that your user has permission to install things
+with it. Sometimes npm and brew compete for control over folders, and it's easy
+to mess up folder permissions when trying to resolve those conflicts. If you
+run into these issues, a clean install of homebrew usually solves them.
+
+## Installing Garage as an editable package
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:62]: So this is the part which is different.

File: docs/user/setting_up_your_development_environment.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+- pipenv:
+
+```
+cd path/to/garage/repo
+pipenv --three
+# --pre required because garage has some dependencies with verion numbers <1.0
+pipenv install --pre -e '.[mujoco,dm_control]'
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:70]: `[all]` includes `mujoco` and `dm_control` (in addition to any other options, except `dev`)

File: docs/user/setting_up_your_development_environment.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+source myenv/bin/activate
+pip uninstall garage  # To ensure no existing install gets in the way
+cd path/to/garage/repo
+pip install -e '.[mujoco,dm_control]'
+pip install -e '.[all,dev]'
+```
+
+----
+
+This page was authored by Hayden Shively
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:96]: haha i don't think i really did anything for this page -- just cite yourself! But I appreciate it.

Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:96]: please put this section in `*italics*`

File: docs/user/setting_up_your_development_environment.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
+# Setting up your development environment
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:1]: perhaps you walk the user through a quick check to make sure everything is working.

possibilities:
* they can run `make docs` to generate the documentation
* they can run `pytest` on some smallish test to make sure they can run tests
* they can run `make test` to verify that Docker is working (though this requires they install docker -- you should note that since it's the primary way of running tests outside of the CI)
* they can edit some file in the repo which prints something on import/construction, and verify that when they run a launcher it prints (as opposed to not printing, which would suggest they have not successfully installed the package as editable).

File: docs/user/setting_up_your_development_environment.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+installed. However, many of the environments garage is used with have additional
+dependencies, and we provide "setup scripts" for installing those dependencies
+and working around known problems on common platforms.
+
+If you can already use the environments you need, you can skip this section.
+
+A MuJoCo key is required for full functionality. You can get one
+[here](https://www.roboti.us/license.html).
+
+### Running the installation scripts
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:34]: you should note that, while optional for regular usage, the dev install has *all* features so this is virtuall required

File: docs/user/setting_up_your_development_environment.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+_Note: Starting with our 2020.06 release, we only support Python 3.6+._
+
+## Installing Dependencies
+
+Generally speaking, system dependencies of garage are minimal, and likely already
+installed. However, many of the environments garage is used with have additional
+dependencies, and we provide "setup scripts" for installing those dependencies
+and working around known problems on common platforms.
+
+If you can already use the environments you need, you can skip this section.
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:29]: this doesn't apply to devs, because they need to run tests

**Comment by haydenshively** [comment]
> Can you move the "extra steps for garage developers" section from `Installation` to here? You can cross-link into the heading here.

Do you mean have the Installation page link to this one, or have this one link to the installation page?

**Comment by ryanjulian** [comment]
I mean that you can cross-link from the installation page to this page.

File: docs/user/setting_up_your_development_environment.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+and working around known problems on common platforms.
+
+Garage's test suite requires *all dependencies* to be installed. Therefore, as a
+developer, you'll need a MuJoCo license key. You can get one
+[here](https://www.roboti.us/license.html).
+
+For information on how to install the dependencies, check out our [Installation
+Guide](installation.rst).
+
+## Installing Garage as an editable package
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:21]: lowercase garage

File: docs/user/setting_up_your_development_environment.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
+# Setting up your development environment
+
+In this section you will learn how to install Garage and its dependencies in a
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:3]: lowercase garage

File: docs/user/setting_up_your_development_environment.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+```
+source myenv/bin/activate
+pip uninstall garage  # To ensure no existing install gets in the way
+cd path/to/garage/repo
+pip install -e '.[all,dev]'
+```
+
+## Verifying that your environment works
+
+Outside of the CI, one of the best ways to test Garage is via Docker. If you
```
Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:52]: lower-case garage

Comment (ryanjulian) [inline docs/user/setting_up_your_development_environment.md:52]: i don't think you should use this example because it requires someone setup Docker. now they need more dependencies.

what's a simple thing they can try?

they can run pytest
they can run pylint
they can build the docs (make docs)

**Comment by ryanjulian** [comment]
@Mergifyio rebase

File: docs/user/setting_up_your_development_environment.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+```
+source myenv/bin/activate
+pip uninstall garage  # To ensure no existing install gets in the way
+cd path/to/garage/repo
+pip install -e '.[all,dev]'
+```
+
+## Verifying that your environment works
+
+As a simple test of your installation, navigate to garage's root directory and
```
Comment (irisliucy) [inline docs/user/setting_up_your_development_environment.md:52]: Testing the garage environment with `make doc` seems irrelevant. I guess to test whether the environment works is by running an example/test case as mentioned below. 

Maybe, you can simply focus on how a user can test on a simple garage example code, say `python examples/tf/vpg_cartpole.py` or `pytest tests/garage/np/algos/test_cem.py` to verify if the environment is working fine.

File: docs/user/installation.rst [hunk 15]
```diff
@@ -98,36 +105,7 @@ pip:
...
-
-- virtualenv
-
-.. code-block:: bash
-
-    source myenv/bin/activate
-    pip uninstall garage  # To ensure no existing install gets in the way.
-    cd path/to/garage/repo
-    pip install -e '.[all,dev]'
+See `here <setting_up_your_development_environment.md#installing-garage-as-an-editable-package>`_
```
Comment (yeukfu) [inline docs/user/installation.rst:108]: This seems not working in the [latest doc](https://garage.readthedocs.io/en/latest/user/installation.html).

Comment (ryanjulian) [reply docs/user/installation.rst:108]: @yeukfu can you file an issue with more details?

Comment (yeukfu) [reply docs/user/installation.rst:108]: Sure.
